zerver/actions/users.py

```diff
@@ -372,11 +378,108 @@ def send_stream_events_for_role_update(
         send_event_on_commit(user_profile.realm, event, [user_profile.id])
 
 
+def unpack_modified_custom_profile_fields(
+    custom_profile_fields: List[ProfileData],
+) -> List[Dict[str, Tuple[str, str]]]:
+    changed_fields: List[Dict[str, Tuple[Any, Any]]] = []
+    old_custom_data: ProfileData = custom_profile_fields[0]
+    new_custom_data: ProfileData = custom_profile_fields[1]
+
+    def replace_integer_literal_with_text_string(cpf_data_element: ProfileDataElement) -> str:
+        cpf_key: str = str(cpf_data_element["value"])
+        cpf_key = ast.literal_eval(cpf_data_element["field_data"])[cpf_key]["text"]
+        return cpf_key
+
+    def retrive_name_from_user_ids(cpf_data_element: ProfileDataElement) -> str:
+        user_names: List[str] = []
+        u_ids = cpf_data_element["value"]
+        if (
+            u_ids is not None
+            and isinstance(u_ids, list)
+            and all(isinstance(val, int) for val in u_ids)
+        ):
+            # To avoid "incompatible type" error during static type checking
+            user_names = [get_user_profile_by_id(u_id).full_name for u_id in u_ids]
+            string_user_names = ", ".join(user_names)
+            if len(user_names) < 2:
+                string_user_names.replace(",", "")
+
+        return string_user_names
+
+    for old_d, new_d in zip(old_custom_data, new_custom_data):
+        # Replace integer literal values with actual user-facing values if there any
+        if old_d["value"] is not None and old_d["value"] in list(string.digits):
+            old_d["value"] = replace_integer_literal_with_text_string(old_d)
+        if new_d["value"] is not None and new_d["value"] in list(string.digits):
+            new_d["value"] = replace_integer_literal_with_text_string(new_d)
+        # In chat.zulip realm some custom_profile_data, i.e. 'Mentor' maps to a list of user_ids
+        if old_d["value"] is not None and isinstance(old_d["value"], list):
+            old_d["value"] = retrive_name_from_user_ids(old_d)
+        if new_d["value"] is not None and isinstance(new_d["value"], list):
+            new_d["value"] = retrive_name_from_user_ids(new_d)
+
+        # unpack the modified values from ProfileData
+        if old_d["value"] != new_d["value"]:
+            changed_fields.append({old_d["name"]: (old_d["value"], new_d["value"])})
+
+    return changed_fields
```

Comment:
Our code style is to not do abbreviations like `new_d` for variable names; they're hard to read.